sw/device/tests/penetrationtests/firmware/fi/otp_fi.c

```diff
@@ -0,0 +1,320 @@
+    uj_output.vendor_test_comp[i] = otp_read32_result_vendor_test_comp[i];
+    uj_output.vendor_test_fi[i] = otp_read32_result_vendor_test_fi[i];
+  }
+  uj_output.otp_status_codes = status.codes;
+  memcpy(uj_output.otp_error_causes, (uint8_t *)status.causes,
+         kDifOtpCtrlStatusCodeHasCauseLast + 1);
+  uj_output.alerts[0] = reg_alerts.alerts[0];
+  uj_output.alerts[1] = reg_alerts.alerts[1];
+  uj_output.alerts[2] = reg_alerts.alerts[2];
+  RESP_OK(ujson_serialize_otp_fi_vendortest_partition_t, uj, &uj_output);
+
+  return OK_STATUS();
+}
+
+status_t handle_otp_fi_owner_sw_cfg(ujson_t *uj) {
+  // Clear registered alerts in alert handler.
+  sca_registered_alerts_t reg_alerts = sca_get_triggered_alerts();
+
+  // Read OTP partition for comparison values
+  TRY(otp_owner_sw_cfg_dump(otp_read32_result_owner_sw_cfg_comp));
+
+  // FI code target.
+  sca_set_trigger_high();
+
+  // Point for FI
+  NOP1000;
+  NOP1000;
+  NOP1000;
+  NOP1000;
+
+  sca_set_trigger_low();
+
+  // Read OTP partition again to see if values changed
+  TRY(otp_owner_sw_cfg_dump(otp_read32_result_owner_sw_cfg_fi));
+
+  // Get registered alerts from alert handler.
+  reg_alerts = sca_get_triggered_alerts();
+
+  // Get OTP CTRL status
+  dif_otp_ctrl_status_t status;
+  TRY(dif_otp_ctrl_get_status(&otp, &status));
+
+  // Send result & status codes to host.
+  otp_fi_ownerswcfg_partition_t uj_output;
+  for (uint32_t i = 0; i < OTP_CTRL_PARAM_OWNER_SW_CFG_SIZE / 4; i++) {
+    uj_output.owner_sw_cfg_comp[i] = otp_read32_result_owner_sw_cfg_comp[i];
+    uj_output.owner_sw_cfg_fi[i] = otp_read32_result_owner_sw_cfg_fi[i];
+  }
+  uj_output.otp_status_codes = status.codes;
+  memcpy(uj_output.otp_error_causes, (uint8_t *)status.causes,
+         kDifOtpCtrlStatusCodeHasCauseLast + 1);
+  uj_output.alerts[0] = reg_alerts.alerts[0];
+  uj_output.alerts[1] = reg_alerts.alerts[1];
+  uj_output.alerts[2] = reg_alerts.alerts[2];
+  RESP_OK(ujson_serialize_otp_fi_ownerswcfg_partition_t, uj, &uj_output);
+
+  return OK_STATUS();
+}
+
+status_t handle_otp_fi_hw_cfg(ujson_t *uj) {
+  // Clear registered alerts in alert handler.
+  sca_registered_alerts_t reg_alerts = sca_get_triggered_alerts();
+
+  // Read OTP partition for comparison values
+  TRY(otp_hw_cfg_dump(otp_read32_result_hw_cfg_comp));
+
+  // FI code target.
+  sca_set_trigger_high();
+
+  // Point for FI
+  NOP1000;
+  NOP1000;
+  NOP1000;
+  NOP1000;
+
+  sca_set_trigger_low();
+
+  // Read OTP partition again to see if values changed
+  TRY(otp_hw_cfg_dump(otp_read32_result_hw_cfg_fi));
+
+  // Get registered alerts from alert handler.
+  reg_alerts = sca_get_triggered_alerts();
+
+  // Get OTP CTRL status
+  dif_otp_ctrl_status_t status;
+  TRY(dif_otp_ctrl_get_status(&otp, &status));
+
+  // Send result & status codes to host.
+  otp_fi_hwcfg_partition_t uj_output;
+  for (uint32_t i = 0; i < OTP_CTRL_PARAM_HW_CFG_SIZE / 4; i++) {
+    uj_output.hw_cfg_comp[i] = otp_read32_result_hw_cfg_comp[i];
```

Comment:
Is there a reason why you are not doing the comparison on the chip instead of sending back all data?

In terms of performance, it could be faster to do the comparison on the chip to avoid the communication overhead?